File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int total;
    public int sumNumbers(TreeNode root) {
        this.total = 0;
        dfs(root, 0);
        return total;
    }
    
    private void dfs(TreeNode root, int cur) {
        if (root == null)   return;
        
        if (root.left == null && root.right == null)
            total += (cur * 10 + root.val);
        
        dfs(root.left, cur * 10 + root.val);
        dfs(root.right, cur * 10 + root.val);
    }
}

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def dfs(root, path = ""):
            if not root:    return
            if not root.left and not root.right:
                nonlocal total_sum
                total_sum += int(path + str(root.val))
                return

            dfs(root.left, path + str(root.val))
            dfs(root.right, path + str(root.val))
        
        total_sum = 0
        dfs(root)
        return total_sum